File: frontend/src/generation/templates/environnement/test_environnement.js
// test_environnement.js
// Tests avancés pour la gestion environnement (sécurité, rôles, plugins, audit, i18n)

/**
 * @file Test environnement (unit, integration, e2e)
 * @description Teste toutes les routes environnement, sécurité, plugins, multitenancy, audit, RGPD.
 * @i18n fr, en, ar, amazigh, de, zh, ja, ko, nl, he, fa, hi, es
 * @roles admin, user, invité
 * @security JWT, CORS, WAF, anti-DDOS
 * @audit log, anonymisation, export
 */

const request = require('supertest');
const app = require('../../../../backend/app');
const { getToken } = require('../../../../backend/utils/testUtils');

describe('Environment API', () => {
  let adminToken, userToken;
  beforeAll(async () => {
    adminToken = await getToken('admin');
    userToken = await getToken('user');
  });

  it('should create an environment dashboard (admin)', async () => {
    const res = await request(app)
      .post('/api/environnement/dashboards')
      .set('Authorization', `Bearer ${adminToken}`)
      .send({
        name: 'Dashboard IA',
        type: 'Monitoring',
        lang: 'fr',
        plugins: ['seo', 'audit'],
      });
    expect(res.statusCode).toBe(201);
    expect(res.body).toHaveProperty('id');
  });

  it('should reject creation for guest', async () => {
    const res = await request(app)
      .post('/api/environnement/dashboards')
      .send({ name: 'Test', type: 'Simple' });
    expect(res.statusCode).toBe(401);
  });

  it('should list dashboards (user)', async () => {
    const res = await request(app)
      .get('/api/environnement/dashboards')
      .set('Authorization', `Bearer ${userToken}`);
    expect(res.statusCode).toBe(200);
    expect(Array.isArray(res.body)).toBe(true);
  });

  it('should update a dashboard (admin)', async () => {
    const create = await request(app)
      .post('/api/environnement/dashboards')
      .set('Authorization', `Bearer ${adminToken}`)
      .send({ name: 'Dashboard à modifier', type: 'Simple' });
    const id = create.body.id;
    const res = await request(app)
      .put(`/api/environnement/dashboards/${id}`)
      .set('Authorization', `Bearer ${adminToken}`)
      .send({ name: 'Dashboard modifié', type: 'Optimisé' });
    expect(res.statusCode).toBe(200);
    expect(res.body.name).toBe('Dashboard modifié');
  });

  it('should delete a dashboard (admin)', async () => {
    const create = await request(app)
      .post('/api/environnement/dashboards')
      .set('Authorization', `Bearer ${adminToken}`)
      .send({ name: 'Dashboard à supprimer', type: 'Simple' });
    const id = create.body.id;
    const res = await request(app)
      .delete(`/api/environnement/dashboards/${id}`)
      .set('Authorization', `Bearer ${adminToken}`);
    expect(res.statusCode).toBe(204);
  });

  it('should log all actions for audit', async () => {
    const res = await request(app)
      .get('/api/environnement/audit')
      .set('Authorization', `Bearer ${adminToken}`);
    expect(res.statusCode).toBe(200);
    expect(res.body).toHaveProperty('logs');
  });
});

File: backend/flask/app/templates/environnement/test_environnement.py
"""
Tests avancés pour le module Environnement (unitaires, intégration, e2e, mocks, sécurité, i18n).
"""
import pytest
from flask import Flask
from flask_jwt_extended import create_access_token
from backend.flask.app.templates.environnement import environnement_bp

@pytest.fixture
def client():
    app = Flask(__name__)
    app.config['JWT_SECRET_KEY'] = 'test'
    app.register_blueprint(environnement_bp)
    with app.test_client() as client:
        yield client

@pytest.fixture
def admin_token():
    return create_access_token(identity={'role': 'admin'})

@pytest.fixture
def user_token():
    return create_access_token(identity={'role': 'user'})

def test_list_projects_admin(client, admin_token):
    res = client.get('/api/environnement/', headers={'Authorization': f'Bearer {admin_token}'})
    assert res.status_code == 200
    assert isinstance(res.json, list)

def test_create_project_admin(client, admin_token):
    res = client.post('/api/environnement/', json={'name': 'Test', 'description': 'Desc'}, headers={'Authorization': f'Bearer {admin_token}'})
    assert res.status_code == 201
    assert 'name' in res.json

def test_create_project_user_forbidden(client, user_token):
    res = client.post('/api/environnement/', json={'name': 'Test', 'description': 'Desc'}, headers={'Authorization': f'Bearer {user_token}'})
    assert res.status_code == 403
